python/ray/util/ml_utils/checkpoint_manager.py

```diff
@@ -0,0 +1,378 @@
+import gc
+import heapq
+import logging
+import numbers
+import os
+import shutil
+
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Optional, Dict, Union, Callable, Tuple, List, Any
+
+import ray
+from ray.tune.result import NODE_IP
+from ray.util import PublicAPI
+from ray.util.annotations import DeveloperAPI
+from ray.util.ml_utils.util import is_nan
+
+MAX = "max"
+MIN = "min"
+
+logger = logging.getLogger(__name__)
+
+
+@DeveloperAPI
+class TrackedCheckpoint:
+    """Checkpoint tracked by a checkpoint manager.
+
+    This class is used to track checkpoints generated by trainables and trainers in
+    order to add metadata (e.g. the result, or the node where it has been created)
+    and for bookkeeping purposes.
+
+    Args:
+        dir_or_data: Checkpoint directory, checkpoint data, or a future to either.
+        storage_mode: Either MEMORY or PERSISTENT.
+        checkpoint_id: Checkpoint number. Usually this should be monotonically
```

Comment:
Any case that this wouldn't be the case? Asking this because this class is a developer API. What would happen if it's not monotonically increasing? Would that bring any surprise?